python/ray/experimental/sgd/tensorflow/tensorflow_trainer.py

```diff
@@ -0,0 +1,201 @@
+                'data_creator', and 'optimizer_creator'.
+            num_replicas (int): the number of workers used in distributed
+                training.
+            use_gpu (bool): Sets resource allocation for workers to 1 GPU
+                if true.
+            batch_size (int): batch size for an update.
+            backend (string): backend used by distributed PyTorch.
+        """
+        self.model_creator = model_creator
+        self.data_creator = data_creator
+        self.config = {} if config is None else config
+        self.batch_size = batch_size
+        self.use_gpu = use_gpu
+        self.num_replicas = num_replicas
+        self.verbose = True
+
+        # Generate actor class
+        Runner = ray.remote(
+            num_cpus=1, num_gpus=int(use_gpu))(TensorFlowRunner)
+
+        if num_replicas == 1:
+            # Start workers
+            self.workers = [
+                Runner.remote(model_creator, data_creator, self.config,
+                              batch_size)
+            ]
+            # Get setup tasks in order to throw errors on failure
+            ray.get(self.workers[0].setup.remote())
+        else:
+            # Start workers
+            self.workers = [
+                Runner.remote(model_creator, data_creator, self.config,
+                              batch_size) for i in range(num_replicas)
+            ]
+
+            # Compute URL for initializing distributed setup
+            ips = ray.get(
+                [worker.get_node_ip.remote() for worker in self.workers])
+            ports = ray.get(
+                [worker.find_free_port.remote() for worker in self.workers])
+
+            urls = [
+                "{ip}:{port}".format(ip=ips[i], port=ports[i])
+                for i in range(len(self.workers))
+            ]
+
+            # Get setup tasks in order to throw errors on failure
+            ray.get([
+                worker.setup_distributed.remote(urls, i, len(self.workers))
+                for i, worker in enumerate(self.workers)
+            ])
+
+    def train(self):
+        """Runs a training epoch."""
+        worker_stats = ray.get([w.step.remote() for w in self.workers])
+        stats = worker_stats[0].copy()
+        return stats
+
+    def validate(self):
+        """Evaluates the model on the validation data set."""
+        stats = ray.get(self.workers[0].validate.remote())
+        return stats
+
+    def get_model(self):
+        """Returns the learned model."""
+        state = ray.get(self.workers[0].get_state.remote())
+        return self._get_model_from_state(state)
+
+    def save(self, checkpoint):
+        """Saves the model at the provided checkpoint.
+
+        Args:
+            checkpoint (str): Path to target checkpoint file.
+
+        """
+
+        state = ray.get(self.workers[0].get_state.remote())
+
+        model = self._get_model_from_state(state)
+
+        model.save(checkpoint + ".h5")
```

Comment:
like technically we just need to dump the “state” to file via pickle; no need to use model. save